### ISSUE TITLE
🐛 BUG: Astro deletes content injecting `<head>` & `<body>`

### ISSUE THREAD

**Issue by jonathantneal** [withastro/astro#2132]
### What version of `astro` are you using?

0.21.10

### What package manager are you using?

npm

### What operating system are you using?

Mac

### Describe the Bug

While investigating https://github.com/withastro/astro/issues/2128, I noticed similar issues when Astro injects `<body>`.

**test-1.astro**: Works!
```astro
---

---
<html lang="en">

<meta charset="utf-8" />
<meta name="viewport" content="width=device-width" />

<title>Test 1</title>

<h1>Test 1</h1>

</html>
```

**test-2.astro**: Fails.
```astro
---
const H1 = 'h1'
---
<html lang="en">

<meta charset="utf-8" />
<meta name="viewport" content="width=device-width" />

<title>Test 1</title>

<H1>Test 1</H1>

</html>
```

**test-2/index.html**: Rendered.
```html
<html lang="en">

<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width">

<title>Test 2</title>

<h1>undefined</h1>
</head><body>Test 2</body>

</html>
```

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-ffn9hf

**Comment by natemoo-re** [comment]
🤔 Looks like these are in `src/pages`, which is subject to our automatic head/body detection heuristics. There are discussions to move away from this behavior and rely on the behavior that occurs outside of `src/pages` and `src/layouts`. Would be very interested to see the output if these files used that behavior.

Not sure we could ever expect `test-2.astro` to work as you expect with the current behavior! How could the compiler differentiate between a component like `MetaHead` that you _mean_ to be in the `<head>` and `H1` that you _mean_ to be in the `<body>`.

**Comment by jonathantneal** [comment]
@natemoo-re, would this be resolved by https://github.com/withastro/astro/pull/2193?

**Comment by tony-sull** [comment]
Looks like this one would likely be part of the improvements for the [RFC](https://github.com/withastro/rfcs/blob/main/proposals/0007-finalize-head-body.md) to finalize `<head>` and `<body>` behavior :+1:

**Comment by natemoo-re** [comment]
Thanks @tony-sull! This is on track to be fixed in https://github.com/withastro/compiler/pull/267 and the subsequent `0.24.x` release of `astro`.

**Comment by natemoo-re** [comment]
This should be fixed in the newly released `astro@0.24.0`! 

Please give feel free to reopen this issue if you're still having trouble.
